### PR TITLE
Sparse Index: integrate with reset

### PR DESCRIPTION
**Pull request by vdye**
This series integrates the sparse index with `git reset` and provides miscellaneous fixes and improvements to the command in sparse checkouts. This includes: 

1. tests added to `t1092` and `p2000` to establish the baseline functionality of the command
2. repository settings to enable the sparse index with `ensure_full_index` guarding any code paths that break tests without other compatibility updates.
3. modifications to remove or reduce the scope in which `ensure_full_index` must be called.

The sparse index updates are predicated on a fix originating from the microsoft/git fork [1], correcting how `git reset --mixed` handles resetting entries outside the sparse checkout definition. Additionally, a performance "bug" in `next_cache_entry` with sparse index is corrected, preventing repeatedly looping over already-searched entries. 

The `p2000` tests demonstrate a ~70% execution time reduction in `git reset` using a sparse index, and no change (within expected variability [2]) using a full index. Results summarized below [3, 4]:

```
Test                           base              [5/8]                 
-----------------------------------------------------------------------
git reset --hard (full-v3)     1.00(0.50+0.39)   0.97(0.50+0.37) -3.0% 
git reset --hard (full-v4)     1.00(0.51+0.38)   0.96(0.50+0.36) -4.0% 
git reset --hard (sparse-v3)   1.68(1.17+0.39)   1.37(0.91+0.35) -18.5%
git reset --hard (sparse-v4)   1.70(1.18+0.40)   1.41(0.94+0.35) -17.1%

Test                           base              [6/8]   
-----------------------------------------------------------------------
git reset --hard (full-v3)     1.00(0.50+0.39)   0.94(0.48+0.34) -6.0% 
git reset --hard (full-v4)     1.00(0.51+0.38)   0.95(0.51+0.34) -5.0% 
git reset --hard (sparse-v3)   1.68(1.17+0.39)   0.46(0.05+0.29) -72.6%
git reset --hard (sparse-v4)   1.70(1.18+0.40)   0.46(0.06+0.29) -72.9%

Test                               base              [7/8]
---------------------------------------------------------------------------
git reset (full-v3)                0.77(0.27+0.37)   0.72(0.26+0.32) -6.5%
git reset (full-v4)                0.75(0.27+0.34)   0.73(0.26+0.32) -2.7%
git reset (sparse-v3)              1.44(0.96+0.36)   0.43(0.04+0.96) -70.1%
git reset (sparse-v4)              1.46(0.97+0.36)   0.43(0.05+0.79) -70.5%
git reset -- missing (full-v3)     0.72(0.26+0.32)   0.69(0.26+0.30) -4.2%
git reset -- missing (full-v4)     0.74(0.28+0.33)   0.71(0.27+0.32) -4.1% 
git reset -- missing (sparse-v3)   1.45(0.97+0.35)   0.81(0.42+0.90) -44.1%
git reset -- missing (sparse-v4)   1.41(0.94+0.34)   0.79(0.42+0.76) -44.0%

Test                               base              [8/8]            
---------------------------------------------------------------------------
git reset -- missing (full-v3)     0.72(0.26+0.32)   0.73(0.26+0.33) +1.4% 
git reset -- missing (full-v4)     0.74(0.28+0.33)   0.74(0.27+0.32) +0.0% 
git reset -- missing (sparse-v3)   1.45(0.97+0.35)   0.43(0.05+0.80) -70.3%
git reset -- missing (sparse-v4)   1.41(0.94+0.34)   0.44(0.05+0.76) -68.8%
```

## Changes since V1
* Add `--force-full-index` option to `update-index`. The option is used circumvent changing `command_requires_full_index` from its default value - right now this is effectively a no-op, but will change once `update-index` is integrated with sparse index. By using this option in the `t1092` expand/collapse test, the command used to test will not need to be updated with subsequent sparse index integrations.
* Update implementation of mixed reset for entries outside sparse checkout definition. The condition in which a file should be checked out before index reset is simplified to "if it has `skip-worktree` enabled and a reset would change the file, check it out". 
  * After checking the behavior of `update_index_from_diff` with renames, found that the diff used by `reset` does not produce diff queue entries with different pathnames for `one` and `two`. Because of this, and that nothing in the implementation seems to rely on identical path names, no `BUG` check is added.
* Correct a bug in the `sparse index is not expanded` tests in `t1092` where failure of a `git reset --mixed` test was not being reported. Test now verifies an appropriate scenario with corrected failure-checking.

## Changes since V2
* Replace patch adding checkouts for `git reset --mixed` with sparse checkout with preserving the `skip-worktree` flag (including a new test for `git reset --mixed` and update to `t1092 - checkout and reset (mixed)`)
* Move rename of `is_missing` into its own patch
* Further extend `t1092` tests and remove unnecessary commands/tests where possible
* Refine logic determining which pathspecs require `ensure_full_index` in `git reset --mixed`, add related `ensure_not_expanded` tests
* Add `index_search_mode` enum to `index_name_stage_pos`
* Clean up variable usage & remove unnecessary `subtree_path` in `prime_cache_tree_rec`
* Update cover letter performance data 
* More thoroughly explain changes in each commit message

## Changes since V3
* Replace `git update-index --force-full-index` with `git reset update-folder1 -- folder1/a`, remove introduction of new `--force-full-index` option entirely, and add comment clarifying the intent of `sparse-index is expanded and converted back` test
* Fix authorship on `reset: preserve skip-worktree bit in mixed reset` (current patch fully replaces original patch, but metadata of the original wasn't properly replaced)

## Changes since V4
* Update `t1092` test 'checkout and reset (mixed)' to explicitly verify differences between sparse and full checkouts

## Changes since V5
* Update `t1092` test 'reset with wildcard pathspec' with more cases and better checks
* Add "special case" wildcard pathspec check when determining whether to expand the index (avoids double-loop over pathspecs & index entries)

## Changes since V6
* Update `t1092` test 'reset with pathspecs outside sparse definition' to verify index contents
* Simplify "special case" verification for wildcard pathspecs (added in V5)

Thanks!
-Victoria

[1] microsoft@6b8a074
[2] https://lore.kernel.org/git/8b9fe3f8-f0e3-4567-b20b-17c92bd1a5c5@github.com/
[3] If a test and/or commit is not mentioned, there is no significant change to performance
[4] Pathspec "does-not-exist" is changed to "missing" to save space in performance report

cc: stolee@gmail.com
cc: gitster@pobox.com
cc: newren@gmail.com
cc: Taylor Blau <me@ttaylorr.com>
cc: Bagas Sanjaya <bagasdotme@gmail.com>
cc: Ævar Arnfjörð Bjarmason <avarab@gmail.com>
cc: Phillip Wood <phillip.wood123@gmail.com>
cc: Emily Shaffer <emilyshaffer@google.com>

### CODE REVIEW

File: cache-tree.c [hunk 1]
```diff
@@ -757,27 +771,49 @@ static void prime_cache_tree_rec(struct repository *r,
...
+			strbuf_add(&subtree_path, entry.path, entry.pathlen);
+			strbuf_addch(&subtree_path, '/');
+
+			/*
+			* If a sparse index is in use, the directory being processed may be
+			* sparse. To confirm that, we can check whether an entry with that
+			* exact name exists in the index. If it does, the created subtree
+			* should be sparse. Otherwise, cache tree expansion should continue
+			* as normal.
+			*/
```
Comment (derrickstolee) [inline cache-tree.c:791]: nit: the spacing on this comment should have an extra space to have the `*` characters lined up vertically. VS Code's tab-in/out feature frequently loses this space.

File: builtin/reset.c [hunk 2]
```diff
@@ -196,16 +196,44 @@ static int read_from_tree(const struct pathspec *pathspec,
 			  int intent_to_add)
 {
 	struct diff_options opt;
+	unsigned int i;
```
Comment (derrickstolee) [inline builtin/reset.c:199]: nit on commit message: similarly here, I would leave back-ticks out of the top sentence:

```
reset: make --mixed sparse-aware
```

(Also: are we using "sparse-aware" versus "sparse aware" consistently? Let's make sure we are consistent within this series, at least.)

File: t/t1092-sparse-checkout-compatibility.sh [hunk 3]
```diff
@@ -568,7 +568,7 @@ test_expect_success 'reset with sparse directory pathspec outside definition' '
 	test_all_match git status --porcelain=v2
 '
 
-test_expect_success 'reset with file pathspec outside sparse definition' '
+test_expect_success 'reset with pathspec match in sparse directory' '
```
Comment (derrickstolee) [inline t/t1092-sparse-checkout-compatibility.sh]: nit: this test shouldn't change names here. Apply this change earlier.

File: builtin/reset.c (hunk 2 above)
Comment (vdye) [reply builtin/reset.c:199]: Removed the backticks from this (and other commits). I've been using "sparse-aware" consistently in this series, as far as I can tell.

Comment (derrickstolee) [reply builtin/reset.c:199]: I must have mis-remembered.

**Comment by derrickstolee** [comment]
@vdye: have you considered taking only the `git reset` changes for your first series, just as a way to try it out? That will also improve the time-to-merge. From there, you could make a case for `checkout-index` and `update-index` as being necessary for `git stash`, so maybe they get grouped with those changes instead.

Just a thought. If you're comfortable going with a big series right away, then ignore me.

**Comment by derrickstolee** [comment]
I changed the title so you didn't forget (it becomes the subject of your cover letter).

**Pull request closed without merging**